Whisper Lmms-eval

stisiTT opened the pull request:
whisper_tt (model=distil-whisper/distil-large-v3)
|        Tasks         |Version|Filter|n-shot|Metric|   |Value |   |Stderr|
|----------------------|-------|------|-----:|------|---|-----:|---|------|
|librispeech_test_other|Yaml   |none  |     0|wer   |↓  |5.4858|±  |   N/A|

# Audio Evaluation Infrastructure PR

## Overview

This PR adds comprehensive audio evaluation support to the tt-inference-server, enabling Whisper evaluation through lmms-eval integration. The changes introduce new audio-specific infrastructure, evaluation workflows, and enhanced server configurations.

## Technical Details

### Audio Evaluation Flow

1. Docker server launches tt-media-server with Whisper model
2. `run_evals.py` detects AUDIO model type and skips PromptClient initialization
3. lmms-eval (whisper_tt class) communicates directly with `/audio/transcriptions` endpoint
4. AudioClient generates benchmark reports with TTFT metrics
5. Dataset transformation applies based on `--audio-eval-dataset` flag

fivanovicTT committed:
feat: init version for whisper benchmark data

fivanovicTT committed:
feat: make return_perf_metrics public

fivanovicTT committed:
feat: retrieve streaming from cli args

fivanovicTT committed:
Merge branch 'dev' into fivanovic/whisper-accuracy-perf-ttft

fivanovicTT committed:
fix: make whisper benchmarks runnable

fivanovicTT committed:
Merge branch 'dev' into fivanovic/whisper-accuracy-perf-ttft

fivanovicTT committed:
feat: make release for whisper

fivanovicTT committed:
fix: address pr comments

fivanovicTT committed:
feat: init implementation for streaming on whisper

fivanovicTT committed:
feat: streaming method update

fivanovicTT committed:
Merge branch 'dev' into fivanovic/whisper-accuracy-perf-ttft

fivanovicTT committed:
chore: revert settings to default version of dev

fivanovicTT committed:
fix: naming of the model

fivanovicTT committed:
fix: address pr comments

fivanovicTT committed:
chore: make streaming off to be default

fivanovicTT committed:
chore: make streaming default to false

stisiTT committed:
refactor: update settings and docker server configuration

- Changed device_ids in settings to use Pydantic Field with alias.
- Updated model_config to populate by name in settings.
- Enhanced run_docker_server to add environment variables for audio and CNN models.
- Adjusted docker command to conditionally mount directories based on model type.

fivanovicTT committed:
fix: allow hf whisper repo name as model in cli

stisiTT committed:
feat: enhance audio evaluation support

- Added support for audio evaluation datasets with a consolidated configuration.
- Introduced a new command-line argument for selecting audio evaluation datasets.
- Updated evaluation and benchmarking scripts to utilize the new audio client.
- Modified the run_evals and run_reports scripts to apply audio dataset transformations.
- Added a new virtual environment setup function for audio evaluations.

fivanovicTT committed:
feat: read as json

stisiTT committed:
Pre ttnn docker image debugging

stisiTT committed:
Merge branch 'fivanovic/whisper-accuracy-perf-ttft' into samt/fivanovic/whisper-accuracy-perf-ttft_samt_cli_work

fivanovicTT committed:
Merge branch 'dev' into fivanovic/whisper-accuracy-perf-ttft

fivanovicTT committed:
feat: eval and benchmark workflow are now separate; for eval we run models

stisiTT committed:
Merge branch 'fivanovic/whisper-accuracy-perf-ttft' into samt/fivanovic/whisper-accuracy-perf-ttft_samt_cli_work

# Conflicts:
#	evals/run_evals.py

stisiTT committed:
Fixed authentication, broken request into 404 now

tstescoTT committed:
add back vLLM --api-key arg so that it doesnt get plaintext logged (#874)

stisiTT committed:
feat: add support for 64-bit audio formats in AudioManager

stisiTT committed:
Merge branch 'dev' into samt/fivanovic/whisper-accuracy-perf-ttft_samt_cli_work

# Conflicts:
#	benchmarking/benchmark_targets/model_performance_reference.json
#	benchmarking/run_benchmarks.py
#	evals/eval_config.py
#	run.py
#	utils/image_client.py
#	workflows/model_spec.py

stisiTT committed:
Refactor JWT handling in run_vllm_api_server.py

- Moved the setting of the encoded JWT as the VLLM_API_KEY environment variable from the main function to the handle_secrets function to improve code organization and avoid logging sensitive information in vLLM arguments.

stisiTT committed:
Merge branch 'dev' into samt/whisper-lmms-evals

# Conflicts:
#	tt-media-server/config/settings.py

stisiTT commented:
### Accuracy Evaluations for distil-large-v3 on n300

| model           | device | task_name              | accuracy_check | score | ratio_to_reference | gpu_reference_score | ratio_to_published | published_score                                                      |
|-----------------|--------|------------------------|----------------|-------|--------------------|---------------------|--------------------|----------------------------------------------------------------------|
| distil-large-v3 | n300   | librispeech_test_other | PASS ✅         | 5.49  | 0.98               | 96.19               | 1.00               | [94.81](https://huggingface.co/spaces/hf-audio/open_asr_leaderboard) |

bgoelTT commented on benchmarking/benchmark_targets/model_performance_reference.json:
```diff
@@ -709,18 +709,6 @@
-                        "ttft_ms": 400
-                    }
-                }
-            }
```
Why is this removed?

stisiTT committed:
Add support for openai/whisper-large-v3 model

- Updated model normalization and argument parsing to include openai/whisper-large-v3.
- Removed deprecated eval configuration for distil-whisper/distil-large-v3.
- Enhanced model specification to support both whisper models in workflows.
- Modified report generation to accommodate different file naming patterns for whisper models.

stisiTT committed:
Enhance report generation and evaluation logic in run_reports.py

- Added logic to handle converted scores for WER metrics more effectively.
- Updated file naming patterns for evaluation results based on model type (audio vs non-audio).
- Improved handling of performance targets for whisper models, ensuring proper lookup under the correct naming conventions.
- Initialized benchmark summary data when no performance targets are found for a device.

bgoelTT commented on utils/image_client.py:
```diff
@@ -290,7 +290,7 @@ def _generate_report(self, status_list: list[SDXLTestStatus], is_image_generate_
             "model": self.model_spec.model_name,
             "device": self.device.name,
-            "timestamp": time.strftime("%Y-%m-%d %H:%M:%S", time.localtime()),
+            "timestamp": time_module.strftime("%Y-%m-%d %H:%M:%S", time_module.localtime()),
```
I don't see this module being mentioned in the changelog - is this a mistake?

bgoelTT commented on workflows/model_spec.py:
```diff
@@ -1688,8 +1688,8 @@ def expand_to_specs(self) -> List["ModelSpec"]:
-                device=DeviceTypes.GALAXY,
-                max_concurrency=32,
+                device=DeviceTypes.N300,
+                max_concurrency=1,
```
We should not overwrite the Galaxy `DeviceModelSpec` but instead add a new n300 entry

bgoelTT commented on workflows/run_docker_server.py:
```diff
@@ -111,6 +112,25 @@ def run_docker_server(model_spec, setup_config, json_fpath):
         "TT_MODEL_SPEC_JSON_PATH": docker_json_fpath,
     }
 
+    # Add environment variables for tt-media-server containers (audio, image, cnn models)
```
```suggestion
    # TODO: Refactor this to set all of these variables inside the ModelSpec json
    # Add environment variables for tt-media-server containers (audio, image, cnn models)
```

We should add a note for our future selves that we need to abstract this into the ModelSpec json + make media server use the ModelSpec json

bgoelTT reviewed (changes requested):
Mostly looks good - my one big concern is that we're crowding the `run.py` namespace with a specific audio-evaluation related param

stisiTT committed:
Merge branch 'dev' into samt/whisper-lmms-evals

# Conflicts:
#	evals/eval_config.py
#	run.py
#	utils/image_client.py
#	workflows/model_spec.py
#	workflows/run_reports.py

stisiTT committed:
Refactor model specifications and remove whisper model display names

- Removed specific whisper model names from valid model options in `run.py`.
- Updated `model_performance_reference.json` to use simplified model naming.
- Cleaned up `ModelSpec` and `ModelSpecTemplate` classes by removing unused display name attributes and related logic for whisper models.

stisiTT committed:
Distilled working, only:Enhance AudioManager initialization and error handling

- Added error logging during AudioManager initialization in worker processes to prevent crashes.
- Implemented lazy loading for the diarization model, with improved error handling for initialization failures.
- Updated environment variable naming in the Docker server workflow to maintain consistency.

stisiTT committed:
Implement audio and CNN-specific setup functions for benchmarking and evaluation

- Added setup functions for audio and CNN environments in `run_benchmarks.py` and `run_evals.py`.
- Updated the main execution flow to call these setup functions based on model type.
- Introduced environment variable configuration functions for Docker containers specific to audio and CNN models in `run_docker_server.py`.
- Enhanced report generation functions to handle audio and CNN-specific data in `run_reports.py`.
- Created virtual environment setup functions for audio and CNN models in `workflow_venvs.py`.

stisiTT committed:
Merge branch 'dev' into samt/whisper-lmms-evals

# Conflicts:
#	tt-media-server/config/settings.py

stisiTT committed:
Refactor evaluation report generation for CNN and AUDIO models

- Updated WER score conversion logic to streamline accuracy calculations.
- Implemented special handling for CNN and AUDIO models to combine evaluation files into a single JSON output.
- Enhanced report generation to include summary information and improved logging for processing evaluation files.

stisiTT committed:
Add galaxy model performance configuration back and fix audio model type check

- Re-Introduced new performance configuration for the galaxy model in `model_performance_reference.json`.
- Updated the model type check in `run_evals.py` to use the correct enum comparison for AUDIO models.

stisiTT committed:
Refactor model type checks to use enum comparisons

- Updated model type checks across multiple files to utilize the ModelType enum for better clarity and consistency.
- Adjusted conditions in `benchmark_config.py`, `run_benchmarks.py`, `run_evals.py`, `run_reports.py`, and `workflow_venvs.py` to replace string comparisons with enum comparisons for AUDIO and CNN models.

stisiTT committed:
Remove audio evaluation dataset configuration and transformation logic

- Deleted the audio evaluation dataset argument from `run.py` and removed related transformation logic from `eval_config.py`, `run_evals.py`, and `run_reports.py`.
- Updated evaluation tasks to consistently use `librispeech_test_other` as the default task name for Whisper models.
- Introduced constants for PCM audio normalization in `audio_manager.py` to improve clarity and maintainability.

stisiTT committed:
Remove special handling for CNN and AUDIO models in report generation

- Deleted the logic for combining evaluation files into a single JSON output specific to CNN and AUDIO models in `run_reports.py`.
- Simplified the report generation process by removing unnecessary file handling and logging related to these model types.

stisiTT committed:
Merge branch 'dev' into samt/whisper-lmms-evals

bgoelTT commented on utils/audio_client.py:
```diff
@@ -0,0 +1,198 @@
+        
+        with open(eval_filename, "w") as f:
+            json.dump(benchmark_data, f, indent=4)
+        print(f"Evaluation data written to: {eval_filename}")
```
Why are evals running benchmarks?

stisiTT replied:
resolved in [commit](https://github.com/tenstorrent/tt-inference-server/commit/ca3a4c3d3231eea33539a52a8279b57457446894)

bgoelTT commented on workflows/run_docker_server.py:
```diff
@@ -31,6 +32,69 @@ def short_uuid():
+        device_ids_str = ",".join(f"({d})" for d in args.device_id)
+    else:
+        # Default to device 0 for single device setups
+        device_ids_str = "(0)"
```
This is the opposite behaviour of what we do for vLLMs. Can we invert this to use all devices by default?

bgoelTT replied:
https://github.com/tenstorrent/tt-inference-server/blob/337c12df17fe8ffd3a6ee57de601004980363afb/tt-media-server/config/constants.py#L98-L109

Yeah for n300 they open 1x1 meshes

stisiTT replied:
Taking this conversation out of scope of this PR and moving it to [this issue](https://github.com/tenstorrent/tt-inference-server/issues/1027) for future discussion.

bgoelTT commented on workflows/workflow_venvs.py:
```diff
@@ -247,6 +281,25 @@ def setup_evals_vision(
+    run_command(
+        f"{uv_exec} pip install --managed-python --python {venv_config.venv_python} "
+        f"'git+https://github.com/bgoelTT/lmms-eval.git@ben/samt/whisper-tt#egg=lmms-eval[audio]' "
+        f"pyjwt==2.7.0 pillow==11.1 qwen_vl_utils jiwer pytest graphviz",
```
I know pyjwt and pillow, but what are the rest of these requirements for?

stisiTT replied:
These might be an overzealous copy and paste. I'll remove them and see if things still work.

stisiTT commented on benchmarking/run_benchmarks.py:
```diff
@@ -45,6 +46,30 @@
+    logger.info(f"Setting up CNN benchmarks for model: {model_spec.model_name}")
+    # CNN-specific benchmark setup can be added here
+    pass
+
```
You're seeing this as a change because I had to decouple audio configs from being conflated in image configs. 

My scope of work was evals for whisper, not benchmarks. I'm unaware of benchmarks configs for whisper.

stisiTT committed:
Merge branch 'dev' into samt/whisper-lmms-evals

# Conflicts:
#	tt-media-server/model_services/audio_service.py
#	tt-media-server/utils/audio_manager.py
#	workflows/model_spec.py
#	workflows/run_reports.py

stisiTT committed:
Remove unused AudioClient logic and refactor AUDIO model evaluation workflow

- Deleted `_read_latest_benchmark_json` and `run_evals` methods from `audio_client.py` as they are no longer used in the evaluation workflow.
- Updated `run_evals.py` to clarify that AUDIO models use `lmms-eval` directly for accuracy evaluation and bypass `AudioClient`.
- Adjusted logging and comments to reflect removal of AUDIO model benchmarking logic.

stisiTT committed:
Remove unused dependencies from audio evals venv setup

Removed qwen_vl_utils, jiwer, pytest, and graphviz from the audio
evaluation virtual environment setup as they are not required for
audio model evaluation workflows.

stisiTT commented on workflows/run_docker_server.py:
```diff
@@ -31,6 +32,69 @@ def short_uuid():
+        "DEVICE": model_spec.device_type.name.lower(),
+        "DEVICE_IDS": device_ids_str,
+        # Disable audio preprocessing by default to avoid HF_TOKEN requirement for basic transcription
+        "ALLOW_AUDIO_PREPROCESSING": "false",
```
~Your wording threw me off, but I think I understand the question here. Looking into this.~

stisiTT committed:
Update `tt_metal_commit` to 2496be4

stisiTT committed:
Merge branch 'dev' into samt/whisper-lmms-evals

# Conflicts:
#	tt-media-server/config/constants.py

bgoelTT reviewed (approved):
Looks great for now thanks Sam

stisiTT committed:
Merge branch 'dev' into samt/whisper-lmms-evals

stisiTT committed:
Update audio eval score configuration and model spec

- Adjusted `gpu_reference_score` values for audio evaluation tasks in `eval_config.py` and added a reference link for `gpu_reference_score_ref`.
- Updated `tt_metal_commit` hash in `model_spec.py` to `6ff23d7`.
- Removed unused `DeviceModelSpec` configuration for N300 devices.

fivanovicTT committed:
feat: align benchmark workflow; other small fixes

idjuricTT committed:
Merge branch 'dev' into samt/whisper-lmms-evals

idjuricTT committed:
Refactor CNNService by removing unnecessary comments

Removed CNN-specific comments and logging initialization.

idjuricTT committed:
Fix DISTIL_WHISPER_LARGE_V3 constant value

stisiTT committed:
Update OPENAI_WHISPER_LARGE_V3 constant to use simplified naming

idjuricTT committed:
Merge branch 'dev' of https://github.com/tenstorrent/tt-inference-server into samt/whisper-lmms-evals

stisiTT committed:
Merge branch 'dev' into samt/whisper-lmms-evals

# Conflicts:
#	workflows/model_spec.py

stisiTT committed:
Merge branch 'dev' into samt/whisper-lmms-evals

# Conflicts:
#	tt-media-server/config/constants.py

stisiTT committed:
Merge branch 'dev' into samt/whisper-lmms-evals

stisiTT committed:
Remove locust directory mount from Docker server configuration

stisiTT committed:
Fix CNN eval JSON parsing error for list-format files

CNN models (e.g., Stable Diffusion) produce eval files in two formats:
- List format from image_client: [{...}]
- Dict format from lmms-eval: {"results": {...}, "configs": {...}}

The code assumed all files were dict format, causing AttributeError when
processing list-format files. Added format detection and separate processing
paths to handle both formats correctly.

- Add separate_files_by_format() to detect JSON structure
- Add process_list_format_eval_files() for list-format processing
- Update evals_generate_report() to route files by format

Audio models unaffected (they only use dict format).

stisiTT committed:
Merge branch 'dev' into samt/whisper-lmms-evals

stisiTT committed:
Merge branch 'dev' into samt/whisper-lmms-evals

# Conflicts:
#	benchmarking/run_benchmarks.py
#	evals/run_evals.py
#	utils/media_clients/image_client.py

fivanovicTT committed:
feat: remove unreachable code; add tsu and accuracy_check to benchmark

tstescoTT committed:
add dataset limit for librispeech for whisper (#1074)

stisiTT committed:
Merge branch 'dev' into samt/whisper-lmms-evals
